frontend: use vite's mechanism for env variables to fill window.envir…

BacLuc opened the pull request:
…onment

That we don't have locally cached environment.js files breaking the development setup. Sadly we still use vue-cli to run our tests, so there the vite mechanism does not work and we have to maintain one additional environment.js.

In production, we compose the environment.js anyway in the configmap, thus we don't have to consider the production case in dev-environment.js.

Now we can also remove the check in index.html.
In a production environment, you have to check yourself that you have all the needed configurations set.

closes #3267

EDIT:
~~Current state: could not get the tests to run... -> converted it to draft again~~

EDIT:
Now turned the PR around:

Step 1:
replace jest with vitest
See the single commits for some quirky fixes that it works.
Also reduces the coverage by 6% in the frontend, and the overall coverage by 30%

Step 2:
Use the vite env variable mechanism to fill environmnet

I will use review comments in the files that we can discuss the different fixes that were needed.

BacLuc committed:
frontend: use vite's mechanism for env variables to fill window.environment

That we don't have locally cached environment.js files breaking the development setup.
Sadly we still use vue-cli to run our tests, so there the vite mechanism does
not work and we have to maintain one additional environment.js.

In production, we compose the environment.js anyway in the configmap, thus we don't
have to consider the production case in dev-environment.js.

Now we can also remove the check in index.html.
In a production environment, you have to check yourself that you have all the needed configurations
set.

closes #3267

usu commented on frontend/src/dev-environment.js:
```diff
@@ -0,0 +1,21 @@
+    VERSION_LINK_TEMPLATE:
+      env.VITE_VERSION_LINK_TEMPLATE ??
+      'https://github.com/ecamp/ecamp3/commit/{version}',
+    FEATURE_DEVELOPER: env.VITE_FEATURE_DEVELOPER ?? true,
```
This needs some boolean conversion, because the environment variables are always coming as string.

```suggestion
    FEATURE_DEVELOPER: (env.VITE_FEATURE_DEVELOPER ?? 'true') === 'true',
```

BacLuc replied:
done

usu commented on frontend/src/main.js:
```diff
@@ -1,3 +1,5 @@
+//getEnv needs to be imported first
+import { getEnv } from '@/dev-environment'
```
```suggestion
import { getEnv } from '@/dev-environment.js'
```

BacLuc replied:
done

usu reviewed (commented):
We also have a `environment.js` file in the `.jest` folder: https://github.com/ecamp/ecamp3/blob/devel/frontend/.jest/environment.js

This is still needed and properly used, right?

BacLuc commented:
> We also have a `environment.js` file in the `.jest` folder: https://github.com/ecamp/ecamp3/blob/devel/frontend/.jest/environment.js
> 
> This is still needed and properly used, right?

 Sadly we still use vue-cli to run our tests, so there the vite mechanism does not work and we have to maintain one additional environment.js.

carlobeltrame commented:
Migrating from jest / vue-cli to vitest shouldn't be too hard. Maybe I'll try that soon

BacLuc commented on frontend/src/App.vue:
```diff
@@ -28,6 +28,7 @@
 import LanguageSwitcher from '@/components/layout/LanguageSwitcher.vue'
 import VueI18n from '@/plugins/i18n'
 import { parseTemplate } from 'url-template'
+import { getEnv } from '@/dev-environment.js'
```
if i did not use the function in some components,
the file (and specially the init of window.environment( was tree shaken and thus removed from the production build.

Maybe someone knows more about how to setup a javascript project properly

BacLuc commented:
> Migrating from jest / vue-cli to vitest shouldn't be too hard. Maybe I'll try that soon

I tried it. They kept the api from jest, but there is not e a drop in replacement for everything. https://github.com/BacLuc/ecamp3/tree/try-vitest

carlobeltrame commented:
I've gotten a lot further in the vitest migration, and am still actively working on this. Just FYI, so we don't do too much parallel work.
As far as I can tell so far, vite *does* have drop-in replacements for most if not all jest mechanisms which we use. Admittedly, some of them aren't [documented very clearly](https://github.com/vitest-dev/vitest/pull/2927) in the docs. In other places, you threw out our much needed jest extensions / customizations without replacing them (e.g. the [snapshot serializer](https://www.npmjs.com/package/jest-serializer-vue-tjw) works out of the box in vitest, and for the canvas mock there is an [adapter package](https://github.com/wobsoriano/vitest-canvas-mock)). And in yet some other places, we had to wait for some [vitest bugfixes](https://github.com/vitest-dev/vitest/issues/740).

BacLuc commented:
Should i
1. continue this PR [👍]
2. Help @carlobeltrame with migrating to vitest (i just need a link to your branch) [🚀]
3. close this PR [👎]

carlobeltrame commented:
My work on top of yours is now here: https://github.com/carlobeltrame/ecamp3/tree/try-vitest
Right now I don't remember how far I got or what was the immediate next problem. But a good start would probably be to rebase the branch and then update all touched npm packages to their latest versions. As mentioned, some of the involved packages also had some bugfixing to do.

BacLuc committed:
frontend: minimal switch to vitest

Not all tests are running

BacLuc committed:
frontend: add .vue to imports for tests

Because vitest cannot resolve the files if there is no extension.
And update snapshots.

BacLuc committed:
frontend: move method debouncedSave to methods where it belongs

Vitest said correctly that this method does not exist.

BacLuc committed:
frontend: fix ApiWrapper.spec.js for vitests

BacLuc committed:
ApiMock.js: replace jest with vi

BacLuc committed:
Admin.spec.js: adapt to vitest

Let the test helpers create the store, just provide a json.
This fixes Error: [vuex] getters should be function but "getters.getLoggedInUser" is {"_meta":{"self":"/users/17d341a80579"}}.

Write the stubs in camel case

And simplify a little.

BacLuc committed:
frontend: delete RichText tests

Because i could not get them to run with vitest.
The error i could not resolve:
 FAIL  src/components/form/api/__tests__/ApiRichtext.spec.js > An ApiRichtext > updates state if value in store is refreshed and has new value
TypeError: tippy is not a function
 ❯ BubbleMenuView.createTooltip node_modules/@tiptap/extension-bubble-menu/src/bubble-menu-plugin.ts:141:18
 ❯ BubbleMenuView.updateHandler node_modules/@tiptap/extension-bubble-menu/src/bubble-menu-plugin.ts:200:10
 ❯ BubbleMenuView.update node_modules/@tiptap/extension-bubble-menu/src/bubble-menu-plugin.ts:170:10
 ❯ EditorView.updatePluginViews node_modules/prosemirror-view/dist/index.js:5262:32
 ❯ EditorView.updateStateInner node_modules/prosemirror-view/dist/index.js:5211:14
 ❯ EditorView.updateState node_modules/prosemirror-view/dist/index.js:5142:14
 ❯ Editor.dispatchTransaction node_modules/@tiptap/core/src/Editor.ts:344:15
 ❯ EditorView.dispatch node_modules/prosemirror-view/dist/index.js:5471:33
 ❯ Object.method [as setContent] node_modules/@tiptap/core/src/CommandManager.ts:42:18
 ❯ VueComponent.value src/components/form/tiptap/TiptapEditor.vue:174:1
    172|       // of the input field
    173|       if (val !== this.html) {
    174|         this.editor.commands.setContent(val)
       | ^
    175|       }
    176|     },

BacLuc committed:
auth.spec.js: reduce time for logout test

router.push takes ca 2 seconds, which is too long for a unit tests.
Thus we mock router.push and router.resolve.
Then store.replace does not work anymore, thus we use vi.importActual
to make sure not the somehow mocked store is used.

BacLuc committed:
frontend: configure coverage for vitest

BacLuc committed:
common: make matchingTranslationKeys.spec.js run directory independent for vitest

Mocking seems a little difficult with vitest.
If we keep the /app, the tests work in the docker container, but not
when running the tests with a local interpreter.

BacLuc commented on common/eslint-local-rules/__tests__/matchingTranslationKeys.spec.js:
```diff
@@ -39,134 +41,133 @@ ruleTester.run('local-rules/matching-translation-keys', ruleInstance, {
       code: '$tc("components.hello.world")',
       options: options,
-      filename: '/app/src/components/hello.js',
+      filename: '/src/components/hello.js',
```
I did not get the packageDirectory mock to work, so i used a simple solution.
Else it works in the docker container, but not when run locally with an interpreter.
ok @carlobeltrame ?

carlobeltrame replied:
As long as the tests work in CI and when run in the way we have it documented [in the wiki](https://github.com/ecamp/ecamp3/wiki/Testing-guide), fine by me

BacLuc commented on frontend/package.json:
```diff
@@ -172,61 +173,6 @@
-    "testURL": "http://localhost/",
-    "coverageDirectory": "data/coverage",
-    "collectCoverage": true,
-    "collectCoverageFrom": [
-      "src/**/*.{js,vue}",
-      "common/**/*.js",
-      "!src/components/print/print-react/generatePdf.js"
-    ],
```
Maybe we need to use this config also for the v8 coverage.
Should i do this in this PR, or can we do this in another PR?

carlobeltrame replied:
I would merge this and then see whether we have any coverage problems, and solve those later.

BacLuc commented on frontend/src/components/form/api/__tests__/ApiRichtext.spec.js:
I did not get these tests to run.

commit message:
 frontend: delete RichText tests

Because i could not get them to run with vitest.
The error i could not resolve:
 FAIL  src/components/form/api/__tests__/ApiRichtext.spec.js > An ApiRichtext > updates state if value in store is refreshed and has new value
TypeError: tippy is not a function
 ❯ BubbleMenuView.createTooltip node_modules/@tiptap/extension-bubble-menu/src/bubble-menu-plugin.ts:141:18
 ❯ BubbleMenuView.updateHandler node_modules/@tiptap/extension-bubble-menu/src/bubble-menu-plugin.ts:200:10
 ❯ BubbleMenuView.update node_modules/@tiptap/extension-bubble-menu/src/bubble-menu-plugin.ts:170:10
 ❯ EditorView.updatePluginViews node_modules/prosemirror-view/dist/index.js:5262:32
 ❯ EditorView.updateStateInner node_modules/prosemirror-view/dist/index.js:5211:14
 ❯ EditorView.updateState node_modules/prosemirror-view/dist/index.js:5142:14
 ❯ Editor.dispatchTransaction node_modules/@tiptap/core/src/Editor.ts:344:15
 ❯ EditorView.dispatch node_modules/prosemirror-view/dist/index.js:5471:33
 ❯ Object.method [as setContent] node_modules/@tiptap/core/src/CommandManager.ts:42:18
 ❯ VueComponent.value src/components/form/tiptap/TiptapEditor.vue:174:1
    172|       // of the input field
    173|       if (val !== this.html) {
    174|         this.editor.commands.setContent(val)
       | ^
    175|       }
    176|     },

carlobeltrame replied:
Alright, let's write new ones later.

BacLuc commented on frontend/src/components/form/api/__tests__/ApiWrapper.spec.js:
```diff
@@ -1,63 +1,45 @@
-// Libraries
+import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest'
```
Here i had problems with the virtualTIme.
It did not work as expected, so i used additional promises and resolvers to control the execution of promises.

BacLuc commented on frontend/src/plugins/store/index.js:
```diff
@@ -41,4 +42,6 @@ class StorePlugin {
 export let apiStore
 export let store
 
+export let storePlugin = new StorePlugin()
```
this is for the auth.spec.js test.
As soon as i started mocking the router, store.replaceState was undefined...?

carlobeltrame replied:
I think I was able to simplify this: b33f67eed04033510d7cc2fcd4b59c7d1f4d6a38

BacLuc commented on frontend/src/views/camp/__tests__/Admin.spec.js:
```diff
@@ -123,5 +103,31 @@ function createCampWithRole(role) {
+  }
+}
+
+function createStubs() {
```
here i had to change the stups to CamelCase

carlobeltrame commented on frontend/src/components/form/api/__tests__/ApiWrapper.spec.js:
```diff
@@ -1,63 +1,45 @@
+  return {
+    ...lodash,
+    debounce: (callback) => new Promise((resolve) => (debounce = resolve)).then(callback),
+    set: lodash.set,
+    get: lodash.get,
```
That shouldn't be necessary after the `...lodash` above..?

BacLuc replied:
i also expected that, but it did not work without setting these to explicitly.
Did you test it and it worked?

carlobeltrame commented on frontend/src/components/form/api/__tests__/ApiWrapper.spec.js:
```diff
@@ -1,63 +1,45 @@
+  const lodash = await importOriginal()
+  return {
+    ...lodash,
+    debounce: (callback) => new Promise((resolve) => (debounce = resolve)).then(callback),
```
Looks very funky... This implementation only supports one debounce at a time, because there is only one global `let debounce`, right? But I guess it is fine for now.

BacLuc replied:
yes, but why would you have 2 debounce before a user interaction/web request/response...?

carlobeltrame replied:
It is possible to write a test where e.g. the user triggers two separate debounced buttons, and check whether everything is executed in the expected order. I think this kind of setup is not supported by this mock.

manuelmeister reviewed (approved):
Worked on my machine ;)
